tests/functional/test_integration.py

```diff
@@ -1,14 +1,66 @@
 import pytest
+import subprocess
+import time
+import yaml
 from ops.model import ActiveStatus
 from pytest_operator.plugin import OpsTest
 
 
 @pytest.mark.abort_on_fail
-async def test_build_and_deploy(ops_test: OpsTest, gunicorn_image, statsd_exporter_image):
+async def test_build_and_deploy_with_psql(
+    ops_test: OpsTest,
+    gunicorn_image,
+    statsd_exporter_image,
+    influx_model_name,
+):
+    result = subprocess.check_output(["juju", "controllers", "--format", "yaml"])
+    controller_name = next(
+        filter(
+            lambda item: item[1]["cloud"] == "localhost",
+            yaml.safe_load(result)["controllers"].items(),
+        )
+    )[0]
+    subprocess.check_output(["juju", "switch", controller_name])
+    subprocess.check_output(["juju", "add-model", influx_model_name])
+    subprocess.check_output(["juju", "deploy", "influxdb"])
+    subprocess.check_output(["juju", "offer", "influxdb:query", "influxoffer"])
     charm = await ops_test.build_charm(".")
     resources = {
         "gunicorn-image": gunicorn_image,
         "statsd-prometheus-exporter-image": statsd_exporter_image,
     }
-    await ops_test.model.deploy(charm, resources=resources)
-    await ops_test.model.wait_for_idle(status=ActiveStatus.name)
+    await ops_test.model.deploy(charm, resources=resources, series="focal")
+    await ops_test.model.deploy("postgresql-k8s")
+    await ops_test.model.add_relation(
+        "postgresql-k8s:db",
+        "gunicorn-k8s:pg",
+    )
+    await ops_test.juju(
+        "relate",
+        "gunicorn-k8s:influxdb",
+        f"{controller_name}:admin/{influx_model_name}.influxoffer",
+        check=True,
+    )
+    await ops_test.model.wait_for_idle(status=ActiveStatus.name, raise_on_error=False)
+    subprocess.check_output(
+        ["juju", "destroy-model", influx_model_name, "--force", "--destroy-storage", "-y"]
+    )
+    subprocess.check_output(["juju", "switch", "k8s-ctrl"])
+
+
+async def test_status(ops_test: OpsTest):
+    assert ops_test.model.applications["gunicorn-k8s"].status == ActiveStatus.name
+    assert ops_test.model.applications["postgresql-k8s"].status == ActiveStatus.name
+
+
+async def test_workload_psql_var(ops_test: OpsTest):
+    app = ops_test.model.applications["gunicorn-k8s"]
+    await app.set_config({"environment": "TEST_ENV_VAR: {{pg.db_uri}}"})
+    config = await app.get_config()
+    assert config["environment"]["value"] == "TEST_ENV_VAR: {{pg.db_uri}}"
+    time.sleep(10)
+    gunicorn_unit = app.units[0]
+    action = await gunicorn_unit.run("curl 127.0.0.1")
```

Comment:
Why do you run the curl in the gunicorn unit ?
Isn't it better to get the IP of the unit and run an http request from the test itself ?
Something like
```
    status = await ops_test.model.get_status()
    unit = list(status.applications[app.name].units)[0]
    address = status["applications"][app.name]["units"][unit]["address"]

    response = requests.get(f"http://{address}")
    assert response.status_code == 200
    assert "TEST_ENV_VAR" in response.content
```

Reply:
Since this charm has the sidecar pattern, you cannot obtain the address with juju unit-get private/public address (that is the hook we should use otherwise). In sidecar charms it is expected to run the commands on localhost.

Reply:
I'm surprised, works well with discourse which is a sidecar charm too
As long as a service is listening on that port in the pod, it's available from the pod IP (the service IP unfortunately is not as there's no way through ops to expose a K8S service yet)